src/VisualStudio/Core/SolutionExplorerShim/HierarchyItemToProjectIdMap.cs

```diff
@@ -43,45 +44,62 @@ public bool TryGetProjectId(IVsHierarchyItem hierarchyItem, string targetFramewo
-                        // The canonical name of a hierarchy item must be unique _within_ an hierarchy, but since we're
-                        // examining multiple hierarchies the canonical name could be the same. Indeed this happens when two
-                        // project files are in the same folder--they both use the full path to the _folder_ as the canonical
-                        // name. To distinguish them we also examine the "regular" name, which will necessarily be different
-                        // if the two projects are in the same folder.
-                        // Note that if a project has been loaded with Lightweight Solution Load it won't even have a
-                        // hierarchy, so we need to check for null first.
-                        if (p.Hierarchy != null
-                            && p.Hierarchy.TryGetCanonicalName((uint)VSConstants.VSITEMID.Root, out string projectCanonicalName)
-                            && p.Hierarchy.TryGetItemName((uint)VSConstants.VSITEMID.Root, out string projectName)
-                            && projectCanonicalName.Equals(nestedCanonicalName, System.StringComparison.OrdinalIgnoreCase)
-                            && projectName.Equals(nestedName))
-                        {
-                            if (targetFrameworkMoniker == null)
-                            {
-                                return true;
-                            }
+                        return p.Hierarchy.TryGetTargetFrameworkMoniker((uint)VSConstants.VSITEMID.Root, out string projectTargetFrameworkMoniker)
+                            && projectTargetFrameworkMoniker.Equals(targetFrameworkMoniker);
+                    }
 
-                            return p.Hierarchy.TryGetTargetFrameworkMoniker((uint)VSConstants.VSITEMID.Root, out string projectTargetFrameworkMoniker)
-                                && projectTargetFrameworkMoniker.Equals(targetFrameworkMoniker);
-                        }
+                    return false;
+                })
+                .ToArray();
 
-                        return false;
-                    })
-                    .SingleOrDefault();
+            // If we have multiple candidates then we might be dealing with Web Application Projects. In this case
+            // there will be one main project plus one project for each open aspx/cshtml/vbhtml file, all with
+            // identical properties on their hierarchies. We can find the main project by excluding everything
+            // with a contained document. We may end up looking at lots of documents, so we do this check only if we
+            // must.
+            if (candidateProjects.Length > 1)
+            {
+                candidateProjects = candidateProjects
+                    .Where(p => !p.GetCurrentDocuments().Any(doc => doc is ContainedDocument))
```

Comment:
This forces evaluation of all of the rest of the projects even if you've found one with no `ContainedDocuments` - I'd expand this into a foreach loop that breaks early.